src/app/clusters/operational-credentials-server/operational-credentials-server.cpp

```diff
@@ -319,6 +320,59 @@ void DoRemoveAllFabrics(intptr_t)
 {
     OpenDefaultPairingWindow(ResetAdmins::kYes);
 }
+
+CHIP_ERROR SendNOCResponse(chip::app::Command * commandObj, EmberAfNOCResponseStatus status, uint8_t index, ByteSpan debug_text)
+{
+    app::CommandPathParams cmdParams = { emberAfCurrentEndpoint(), /* group id */ 0, ZCL_OPERATIONAL_CREDENTIALS_CLUSTER_ID,
+                                         ZCL_NOC_RESPONSE_COMMAND_ID, (chip::app::CommandPathFlags::kEndpointIdValid) };
+    TLV::TLVWriter * writer          = nullptr;
+
+    VerifyOrReturnError(commandObj != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    ReturnErrorOnFailure(commandObj->PrepareCommand(cmdParams));
+    writer = commandObj->GetCommandDataElementTLVWriter();
+    ReturnErrorOnFailure(writer->Put(TLV::ContextTag(0), status));
+    if (status == EMBER_ZCL_NOC_RESPONSE_STATUS_SUCCESS)
+    {
+        ReturnErrorOnFailure(writer->Put(TLV::ContextTag(1), index));
+    }
+    ReturnErrorOnFailure(writer->Put(TLV::ContextTag(2), debug_text));
+    ReturnErrorOnFailure(commandObj->FinishCommand());
+
+    return CHIP_NO_ERROR;
+}
+
+EmberAfNOCResponseStatus ConvertToNOCResponseStatus(CHIP_ERROR err)
+{
+    switch (err)
+    {
+    case CHIP_NO_ERROR:
+        return EMBER_ZCL_NOC_RESPONSE_STATUS_SUCCESS;
+    case CHIP_ERROR_INVALID_PUBLIC_KEY:
+        return EMBER_ZCL_NOC_RESPONSE_STATUS_INVALID_PUBLIC_KEY;
+    case CHIP_ERROR_WRONG_NODE_ID:
+        return EMBER_ZCL_NOC_RESPONSE_STATUS_INVALID_NODE_OP_ID;
+    case CHIP_ERROR_CA_CERT_NOT_FOUND:
+    case CHIP_ERROR_CERT_PATH_LEN_CONSTRAINT_EXCEEDED:
+    case CHIP_ERROR_CERT_PATH_TOO_LONG:
+    case CHIP_ERROR_CERT_USAGE_NOT_ALLOWED:
+    case CHIP_ERROR_CERT_EXPIRED:
+    case CHIP_ERROR_CERT_NOT_VALID_YET:
+    case CHIP_ERROR_UNSUPPORTED_CERT_FORMAT:
+    case CHIP_ERROR_UNSUPPORTED_ELLIPTIC_CURVE:
+    case CHIP_ERROR_CERT_LOAD_FAILED:
+    case CHIP_ERROR_CERT_NOT_TRUSTED:
+    case CHIP_ERROR_WRONG_CERT_SUBJECT:
+        return EMBER_ZCL_NOC_RESPONSE_STATUS_INVALID_NOC;
+    case CHIP_ERROR_NO_MEMORY:
+        return EMBER_ZCL_NOC_RESPONSE_STATUS_TABLE_FULL;
+    case CHIP_ERROR_INVALID_FABRIC_ID:
+        return EMBER_ZCL_NOC_RESPONSE_STATUS_INVALID_FABRIC_INDEX;
```

Comment:
This error mapping is not correct.

INVALID_FABRIC_INDEX is only used in `RemoveFabric` when the fabric index used is wrong.

If an NOC has an incorrect or missing fabric ID within, or the fabric ID mismatches between ICAC (if present) and NOC, then `EMBER_ZCL_NOC_RESPONSE_STATUS_INVALID_NODE_OP_ID` (mapping to `InvalidNodeOpId` in spec) should be used.